### PR TITLE
Issue #2551 changes

### PR DESCRIPTION
**Pull request by swaqvalley**
For Issue #2551 created by @o01eg. These changes prompt the user for confirmation when leaving a multiplayer game without setting their status to ready.

### CODE REVIEW

**Comment by alleryn** [comment]
You should probably try to squash/fixup those commits (see https://help.github.com/en/articles/using-git-rebase-on-the-command-line).

That should work at least for the bottom two commits.  Something like:
`git rebase -i HEAD~3`
```
pick 6ae7457 update
pick 3981e1c Prompt user when leaving multiplayer game without setting status to ready.
fixup 49b18145c
```
That first update shouldn't really be in there.  I'm not totally sure how to get rid of that. 
@dbenage-cx could you help us with this?

**Comment by swaqvalley** [comment]
Yeah, I'm still trying to figure out GitHub. I was wondering if I was supposed to delete my fork and recreate it to avoid the merge commit.

**Comment by swaqvalley** [comment]
I was able to combine the two commits with actual changes. No idea how to go about removing the merge commit. If it's a problem I can start over my fork from scratch and do a new pull request.

**Comment by geoffthemedio** [comment]
Rebase you branch on master and then force-push.

**Comment by swaqvalley** [comment]
I tried that but now it says "This branch is 14 commits ahead, 17 commits behind freeorion:master."

**Comment by swaqvalley** [comment]
I think I did it! I followed an example for how to force a fork to be even with the upstream then rebased the branch and it appears to just have the one commit now.

File: default/stringtables/en.txt [hunk 1]
```diff
@@ -2579,6 +2579,9 @@ The Save Game button is disabled, either because this is a multiplayer game in w
 GAME_MENU_REALLY_CONCEDE
 Confirm conceding? This empire will be removed from the game.
 
+GAME_MENU_CONFIRM_NOT_READY
+Confirm quit in unready state?
+
```
Comment (alleryn) [inline default/stringtables/en.txt]: I would maybe suggest rewording this.  "Unready" seems a little ambiguous to me.

How would you feel about "Confirm quit?  Orders are not finalized."  Or something like that?

Comment (swaqvalley) [reply default/stringtables/en.txt]: As far as I am aware you don't need to click the turn button to finalize or submit your orders, that happens automatically. So that make the purpose of clicking the button to indicate to the server that you are ready for the turn to advance so that it can proceed when all the players are ready without waiting for the timer to expire.

That said I don't think my wording is ideal, I was just trying to be concise. The name of that feature is not really given in the game. Maybe the closest thing is "waiting" which is used in the code for the enum and the text in the messages box "Waiting for Other Players".

Some other possible options:
"Confirm quit? You are not waiting."
"Confirm quit? You are not waiting for turn advance."
"Confirm quit? Waiting status not set."
"Confirm quit? Ready status not set."
"Confirm quit without setting ready status?"

Comment (alleryn) [reply default/stringtables/en.txt]: Oh, i thought you did need to click the turn button for your orders to process, but that was just an assumption on my part.  My mistake.

Maybe just something to indicate what the player isn't ready for?
"Confirm quit?  Still revising orders."
"Confirm quit?  Turn advance not set."
"Confirm quit?  Still revising orders, turn advance not set."
"Confirm quit without setting readiness for next turn?"

Comment (swaqvalley) [reply default/stringtables/en.txt]: I like "Confirm quit? Still revising orders." That matches up with how the button changes when you click it which would help indicate what the user needs to do.

Comment (alleryn) [reply default/stringtables/en.txt]: I'm happy with that.

Comment (swaqvalley) [reply default/stringtables/en.txt]: String has been updated.

File: default/stringtables/en.txt [hunk 2]
```diff
@@ -2579,6 +2579,9 @@ The Save Game button is disabled, either because this is a multiplayer game in w
 GAME_MENU_REALLY_CONCEDE
 Confirm conceding? This empire will be removed from the game.
 
+GAME_MENU_CONFIRM_NOT_READY
+Confirm quit? Still revising orders.
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]: This message is also displayed if orders have never been sent for the current turn, not just if they were sent and then the revise orders button was clicked.

Comment (swaqvalley) [reply default/stringtables/en.txt]: True. Would you prefer different wording?

Comment (geoffthemedio) [reply default/stringtables/en.txt]: Ideally it would check whether orders have ever been sent for the turn currently being played, or if there are orders issued but not sent since the revise orders button was clicked, or of no orders have ever been sent for the current turn, or if orders were sent and revise clicked but no no additional orders were then issued, and adjust the message accordingly. Alternatively, in a multiplayer game it could say that the server is waiting for orders, or in single player it could say orders have not yet been sent. Or in all cases it could say orders have not been sent... even though if revising then orders were sent (and just the revisions haven't been sent)...

Comment (alleryn) [reply default/stringtables/en.txt]: This kind of brings me back to my original suggestion:
"Confirm quit?  Orders have not been finalized."
This appears to cover both the case in which orders have never been sent for the current turn, and the case in which orders were sent but are currently in the state of revision.

The objection was that you do not need to click the End Turn button to finalize or submit your orders.  But after thinking it through, while your orders may finalize and submit if you allow the timer to expire, it is still accurate to say that the orders have not been _finalized_ if they are currently in the unsubmitted state or the revising state.  (I think)

This does not cover this case:

>  in single player it could say orders have not yet been sent

but i think that is outside the scope of this PR and related issue, which pertain only to multiplayer.  I'm not sure a confirmation is really needed for singleplayer,  where exiting without advancing the turn is (i think) more "standard" behavior than it is in a situation with multiplayer turns where the player has periods of "forced hiatus".

Comment (swaqvalley) [reply default/stringtables/en.txt]: The message will not be displayed in a single player game.

I would prefer a single wording for the dialog than trying to check whether the user has submitted then cancelled to start revising. It would get extra confusing when a user logs back in to the multiplayer server and it automatically starts revising. So a consistent message regardless of that state would be easier and probably better for the user to understand.

I am also fine with alleryn's suggestion "Confirm quit? Orders have not been finalized." Or if we wanted a more concise version: "Confirm quit? Orders not finalized."

Thoughts?

Comment (geoffthemedio) [reply default/stringtables/en.txt]: > The message will not be displayed in a single player game.

Are you sure / is that true now? I can't test immediately but I think I got the message when reseigning or closing the window in a single player game...

Comment (alleryn) [reply default/stringtables/en.txt]: > > The message will not be displayed in a single player game.
> 
> Are you sure / is that true now? I can't test immediately but I think I got the message when reseigning or closing the window in a single player game...

I'm unable to reproduce this.  I only get the dialog box when i try to exit a multiplayer game.  Exiting a singleplayer game is as before.

I think the && !m_single_player_game is intended to prevent this.

While testing i did produce an error however; i'll comment below.

**Comment by alleryn** [comment]
I found an error while testing:

Reproduce:
* Start client.  Host multiplayer game.
* Start second instance of client.  Join multiplayer game at localhost.
* Accept and start game in the clients.
* In first client, Resign and confirm quit.
* This causes the second client to receive message no longer connected to server.  (Should probably exit to main menu at this point).
* Second client also receives confirm quit orders not ready dialog.
* Cancel quit in second client.
* Second client receives connection to server lost.
* Resign and confirm quit in second client.
* Second client receives 'save game in progress' and hangs.

The first time i found this bug, the client completely hung.  Following these steps, i can still press the quit to main menu without saving from the save dialog.  I'm trying to reproduce the worse crash.

Edit: I've been unable to reproduce the complete hang, but the bug is still there: the client should not remain open after it disconnects from the server.  (At least i think that is desired behavior).

**Comment by swaqvalley** [comment]
Thanks for testing, alleryn. I was doing all my testing with a persistent server so I didn't run into this issue.

I can reproduce the save game in progress hang by reversing which player quits first. If the player hosting the game quits second they get the save window. But I found that it is a soft hang, as cancelling the save drops back to the main menu.

What is the preferred behavior here? Just kick the rest of the players out to the main menu whenever one person quits?

File: default/stringtables/en.txt (hunk 2 above)
Comment (swaqvalley) [reply default/stringtables/en.txt]: Geoff, I have never had the message come up for a single player game in my tests. As alleryn stated, I explicitly check that it is not a single player game before showing the dialog box.

**Comment by swaqvalley** [comment]
Okay, it looks like I can use the skip_savegame flag to determine when the reason for reset/exit is because of an error like server disconnect.

File: default/stringtables/en.txt (hunk 2 above)
Comment (geoffthemedio) [reply default/stringtables/en.txt]: Possibly I tested with a nominally multiplayer game with only one human player, which does produce the popup.

But I think the multiplayer/singleplayer distinction isn't important really...

Regardless, "Confirm quit? Orders not finalized." seems OK if you don't want to adjust for the exactly turn-send-revise-resend state.

**Comment by geoffthemedio** [comment]
I think ideally it would do an autosave whenever quitting, without expecting save info from not-connected players, but that's probably beyond the scope of this pull request...

**Comment by swaqvalley** [comment]
I agree that changing the behavior of the autosave is outside of the scope for this PR.

I believe the commit 019f24f addresses/matches the latest discussion here. Is there anything you are waiting on me for?

**Comment by geoffthemedio** [comment]
This also shows the popup for observers... but probably shouldn't.

**Comment by swaqvalley** [comment]
Ah, that is definitely an oversight. That shouldn't be too hard to fix. I assume I should make a new pull request to handle that?

**Comment by geoffthemedio** [comment]
yes
